lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOIndexSpec.hs

```diff
@@ -0,0 +1,531 @@
+        it "prop_toList_fromSequence" $
+            property prop_toList_fromSequence
+
+    parallel $ describe "Modification" $ do
+
+        it "prop_delete_balance" $
+            property prop_delete_balance
+        it "prop_delete_lookup" $
+            property prop_delete_lookup
+        it "prop_delete_size" $
+            property prop_delete_size
+        it "prop_insert_assets" $
+            property prop_insert_assets
+        it "prop_insert_balance" $
+            property prop_insert_balance
+        it "prop_insert_delete" $
+            property prop_insert_delete
+        it "prop_insert_lookup" $
+            property prop_insert_lookup
+        it "prop_insert_lookup" $
+            property prop_insert_size
+
+    parallel $ describe "Index Selection" $ do
+
+        it "prop_selectRandom_empty" $
+            property prop_selectRandom_empty
+        it "prop_selectRandom_singleton" $
+            property prop_selectRandom_singleton
+        it "prop_selectRandom_one_any" $
+            property prop_selectRandom_one_any
+        it "prop_selectRandom_one_withAsset" $
+            property prop_selectRandom_one_withAsset
+        it "prop_selectRandom_all_any" $
+            property prop_selectRandom_all_any
+        it "prop_selectRandom_all_withAsset" $
+            property prop_selectRandom_all_withAsset
+
+    parallel $ describe "Set Selection" $ do
+
+        it "prop_selectRandomSetMember_empty" $
+            property prop_selectRandomSetMember_empty
+        it "prop_selectRandomSetMember_singleton" $
+            property prop_selectRandomSetMember_singleton
+        it "prop_selectRandomSetMember_coversRangeUniformly" $
+            property prop_selectRandomSetMember_coversRangeUniformly
+
+--------------------------------------------------------------------------------
+-- Invariant properties
+--------------------------------------------------------------------------------
+
+invariantHolds :: UTxOIndex -> Property
+invariantHolds u = checkInvariant u === InvariantHolds
+
+prop_arbitrary_invariant :: UTxOIndex -> Property
+prop_arbitrary_invariant = invariantHolds
+
+prop_shrink_invariant :: UTxOIndex -> Property
+prop_shrink_invariant = conjoin . fmap invariantHolds . shrink
+
+prop_empty_invariant :: Property
+prop_empty_invariant = invariantHolds UTxOIndex.empty
+
+prop_singleton_invariant :: TxIn -> TxOut -> Property
+prop_singleton_invariant i o = invariantHolds $ UTxOIndex.singleton i o
+
+prop_fromSequence_invariant :: [(TxIn, TxOut)] -> Property
+prop_fromSequence_invariant = invariantHolds . UTxOIndex.fromSequence
+
+prop_insert_invariant :: TxIn -> TxOut -> UTxOIndex -> Property
+prop_insert_invariant i o u = invariantHolds $ UTxOIndex.insert i o u
+
+prop_delete_invariant :: TxIn -> UTxOIndex -> Property
+prop_delete_invariant i u = invariantHolds $ UTxOIndex.delete i u
+
+prop_selectRandom_invariant :: UTxOIndex -> SelectionFilter -> Property
+prop_selectRandom_invariant i f = monadicIO $ do
```

Comment:
> Can we try avoid IO and use DRG gen => MonadRandom (MonadPseudoRandom gen) instead? I also see this in many other places.

I agree that it would be nicer to avoid `IO` here.

Just for the record: we held a review call and agreed that it would be acceptable to refactor this in a future PR.